### ISSUE TITLE
How to get an molnet dataset as pandas dataframe?

### ISSUE THREAD

**Issue by schinto** [deepchem/deepchem#1657]
### Pandas dataframes from molnet datasets

I am new to deepchem and use the following code snippet to get pandas dataframe for a molnet dataset.

- Is there a simpler way to get the dataset as Pandas dataframe and to export the full dataset in CSV format with column labels?
- How to get the column labels of the features to the dataframe?
---
```python
import os
import numpy as np
import pandas as pd
from deepchem.molnet import load_delaney

delaney_tasks, delaney_datasets, transformers = load_delaney(featurizer='ECFP', split='random')
train_dataset, valid_dataset, test_dataset = delaney_datasets

def convert_dataset_to_df(dataset):
    col_y = dataset.tasks
    num_col_w = dataset.w.shape[1]
    col_w = [f'task_weight_{i:03}' for i in range(1, num_col_w + 1)]
    num_col_x = dataset.X.shape[1]
    col_x = [f'ECFP4_{i:04}' for i in range(1, num_col_x + 1)]
    columns = col_y + col_w + col_x
    df = pd.DataFrame(np.concatenate((dataset.y, dataset.w, dataset.X),
                                     axis=1),
                      index=dataset.ids,
                      columns=columns)
    df.index.name = 'ID'
    # Downcast fingerprint floats to integer
    df[col_x] = df[col_x].apply(pd.to_numeric, downcast='integer')
    return df
	
train_df = convert_dataset_to_df(train_dataset)
valid_df = convert_dataset_to_df(valid_dataset)
test_df = convert_dataset_to_df(test_dataset)

path = './datasets-prepared'
train_df.to_csv(os.path.join(path, 'delaney-ECFP4-train.csv'), index_label='ID')
valid_df.to_csv(os.path.join(path, 'delaney-ECFP4-valid.csv'), index_label='ID')
test_df.to_csv(os.path.join(path, 'delaney-ECFP4-test.csv'), index_label='ID')
```

Thanks

**Comment by VIGS25** [comment]
Hi, 

> Is there a simpler way to get the dataset as Pandas dataframe and to export the full dataset in CSV format with column labels?

To the best of my knowledge, there is no functionality within DeepChem to turn `deepchem.Dataset` to a `pandas.DataFrame`, just because all models are designed to work with `deepchem.Dataset`, and it offers other benefits while writing and reading from disk. Most datasets in their raw form are available as `.csv` files so you should be able to download them and then just featurize the `SMILES` string and do the splits on that. 

> How to get the column labels of the features to the dataframe?

I'm not sure I follow your question, but in a normal `pandas.DataFrame`, you can get the column labels by simply using `list(df)`. If you meant something else, I'd be happy to help.

**Comment by rbharath** [comment]
This might be a valuable feature to add. That is, a standard method for converting a deepchem dataset to a pandas dataframe. I think we could do it by creating columns for `X`,`y`,`w`, `tasks`.

**Comment by rbharath** [comment]
Here's some pertinent discussion: https://forum.deepchem.io/t/how-to-get-scaffoldsplitter-output-as-a-csv-file/122

Another thought is that we might want to make it possible for a dataset to store per-datapoint metadata. Oftentimes, real world molecule data has a whole bunch of additional metadata attached to it (time of synthesis, time of assay, etc etc). Allowing the dataset to store this would mean the dataset could be correctly transferred from an input CSV/dataframe to a `Dataset` object.

I think riffing off the last answer, it would make sense to add a general `Dataset.to_dataframe()` method that turns an arbitrary dataset into a dataframe.

**Comment by peastman** [comment]
I'd like to go ahead and implement this.  `Dataset.to_dataframe()` can be trivial.  It just takes the numpy arrays for `X`, `y`, `w`, and `ids` and wraps them in a dataframe.

I'd also like to add a static method `Dataset.from_dataframe()` for going the other way.  You would give it a dataframe and the names of the columns to use, and it would return a NumpyDataset from them.

Can you think of any reason to make it any more complicated than that?

**Comment by rbharath** [comment]
Great, thanks for taking this on!

One issue I've heard from users is sometimes they have extra metadata in their source CSV files. They convert to a `Dataset`, apply a `Splitter`, but then have to figure out how to bring along their metadata to their new train/test/validate datasets. It might be worth adding a `Dataset.metadata` field that stores extra per-datapoint metadata. Then `Dataset.from_dataframe` could accept a `metadata` argument that converts additional columns into metadata in the dataset. This might be a more complex change though, and is perhaps beter handled in a separate PR from the `Dataset.to_dataframe()`, `Dataset.from_dataframe()` methods PR.

**Comment by peastman** [comment]
Let's make that a separate PR.  It will need a bit of design of its own.
